go/test/endtoend/throttler/util.go

```diff
@@ -115,6 +119,129 @@ func UpdateThrottlerTopoConfig(clusterInstance *cluster.LocalProcessCluster, ena
+			return nil
+		}
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("timed out waiting for GetSrvKeyspaceNames to contain '%v'", keyspace)
+		case <-ticker.C:
+		}
+	}
+}
+
+// throttleAppRaw runs vtctlclient UpdateThrottlerConfig with --throttle-app flags
+// This retries the command until it succeeds or times out as the
+// SrvKeyspace record may not yet exist for a newly created
+// Keyspace that is still initializing before it becomes serving.
+func throttleAppRaw(vtctldProcess *cluster.VtctldClientProcess, keyspaceName string, throttlerApp throttlerapp.Name, throttle bool) (result string, err error) {
+	args := []string{}
+	args = append(args, "UpdateThrottlerConfig")
+	args = append(args, "--throttle-app", throttlerApp.String())
+	args = append(args, "--throttle-app-duration")
+	if throttle {
+		args = append(args, "1h")
+	} else {
+		args = append(args, "-1h")
+	}
+	args = append(args, keyspaceName)
+
+	ctx, cancel := context.WithTimeout(context.Background(), ConfigTimeout)
+	defer cancel()
+
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for {
+		result, err = vtctldProcess.ExecuteCommandWithOutput(args...)
+		if err == nil {
+			return result, nil
+		}
+		select {
+		case <-ctx.Done():
+			return "", fmt.Errorf("timed out waiting for UpdateThrottlerConfig to succeed after %v; last seen value: %+v, error: %v", ConfigTimeout, result, err)
+		case <-ticker.C:
+		}
+	}
+}
+
+// throttleApp throttles or unthrottles an app
+func throttleApp(clusterInstance *cluster.LocalProcessCluster, throttlerApp throttlerapp.Name, throttle bool) (string, error) {
+	rec := concurrency.AllErrorRecorder{}
+	var (
+		err error
+		res strings.Builder
+	)
+	for _, ks := range clusterInstance.Keyspaces {
+		ires, err := throttleAppRaw(&clusterInstance.VtctldClientProcess, ks.Name, throttlerApp, throttle)
+		if err != nil {
+			rec.RecordError(err)
+		}
+		res.WriteString(ires)
+	}
+	if rec.HasErrors() {
+		err = rec.Error()
+	}
+	return res.String(), err
+}
+
+// ThrottleApp throttles given app name for the next hour
+func ThrottleApp(clusterInstance *cluster.LocalProcessCluster, throttlerApp throttlerapp.Name) (string, error) {
+	return throttleApp(clusterInstance, throttlerApp, true)
+}
+
+// ThrottleApp unthrottles given app name
+func UnthrottleApp(clusterInstance *cluster.LocalProcessCluster, throttlerApp throttlerapp.Name) (string, error) {
+	return throttleApp(clusterInstance, throttlerApp, false)
+}
+
+// ThrottleAppAndWaitUntilTabletsConfirm
+func ThrottleAppAndWaitUntilTabletsConfirm(t *testing.T, clusterInstance *cluster.LocalProcessCluster, throttlerApp throttlerapp.Name) (string, error) {
+	res, err := throttleApp(clusterInstance, throttlerApp, true)
+	if err != nil {
+		return res, err
+	}
+	for _, ks := range clusterInstance.Keyspaces {
+		for _, shard := range ks.Shards {
+			for _, tablet := range shard.Vttablets {
+				WaitForThrottledApp(t, tablet, throttlerApp, true, ConfigTimeout)
```

Comment:
Might be worth doing each of these in a goroutine. Although I don't think we have enough shards and tablets in the tests today that it should matter too much.

Reply:
Yeah, the good news is that those tablets do, meanwhile, update their configs concurrently. Say we had to wait full `42sec` for the first tablet; chances are that by the time we check the next tablets, they will have already picked the change. So this loop really just waits for the slowest tablet, which is the same as if we'd run goroutines. We just have a subsecond overhead for serially checking those successful tablets.